sage.py

```diff
@@ -481,483 +449,311 @@ def getepisodes(action_seq, mcat, plot, debug=False):
-        
-        host_alerts = dict()
-
-        for ev in team:
-            #print(ev[0])
-            h = ev[7]
-            s = ev[1]
-            d = ev[3]
-            c = ev[9]
-            ts = ev[8]
-            sp = ev[2] if ev[2] != None else 65000 
-            dp = ev[4] if ev[4] != None else 65000
-            signature = ev[5]
-            # Simply respect the source,dst format! (Correction: source is always source and dest alwyas dest!)
-            
-            source, dest, port = -1, -1, -1
-            #print(s, d, sp, dp)
-            #assert sp >= dp
-            
-            source = s #if s not in inv_hostip.keys() else inv_hostip[s]
-            dest = d #if d not in inv_hostip.keys() else inv_hostip[d]
-            # explicit name if cant resolve
-            #port = str(dp) if (dp not in port_services.keys() or port_services[dp] == 'unknown') else port_services[dp]['name'] 
-            
-            # say unknown if cant resolve it
-            port = 'unknown' if (dp not in port_services.keys() or port_services[dp] == 'unknown') else port_services[dp]['name']
-            
-            if (source,dest) not in host_alerts.keys() and (dest,source) not in host_alerts.keys():
-                host_alerts[(source,dest)] = []
-                #print(tid, (source,dest), 'first', ev[8])
-                s_t[str(tid)+"|"+str(source)+"->"+str(dest)] = ev[8]
-                
-            if((source,dest) in host_alerts.keys()):
-                host_alerts[(source,dest)].append((dest, c, ts, port, signature)) # TODO: remove the redundant host names
-                #print(source, dest, (micro[c].split('.'))[-1], port)
-            else:
-                host_alerts[(dest,source)].append((source, c, ts, port, signature))
-                #print(dest,source, (micro[c].split('.'))[-1], port)
+    return episodes
 
-        team_data[tid] = host_alerts.items()
-        
-    # Calculate number of alerts over time for each attacker 
-    #print(len(s_t))
-    team_episodes = []
 
+def _create_episode(alert_seq_epi, mcat, tid):
+    # Flatten relevant data from the windows of the corresponding alert sequence
+    services = [alert[3] for window in alert_seq_epi for alert in window]
+    unique_signatures = list(set([alert[4] for window in alert_seq_epi for alert in window]))
+    events = [len(window) for window in alert_seq_epi]
+    alert_volume = round(sum(events) / float(len(events)), 1)
+
+    # Make exact start/end times based on alert timestamps
+    timestamps = [alert[2] for window in alert_seq_epi for alert in window]
+    first_ts, last_ts = min(timestamps), max(timestamps)
+
+    # Make the start/end times the actual elapsed times
+    start_time = (first_ts - start_times[tid]).total_seconds()
+    end_time = (last_ts - start_times[tid]).total_seconds()
+    period = end_time - start_time
+
+    episode = (start_time, end_time, mcat, len(events), alert_volume, period, services, unique_signatures, (first_ts, last_ts))
```

Comment:
In the episode definition, there was `len(raw_events)`, however in the code there was just `events` (or `x[3]`). The value is not used anywhere else, but should it actually be `len(raw_events)`?

Reply:
No, events is good enough. Len() can be computed later too, as needed, but the events cant be extracted if only len() is stored.

Reply:
Right, the `events` list is a list of lengths of the alerts windows in the alert sequence for this episode, so it is essentially the length, and this is probably what was meant in the episode definition. My bad